Define __signature__ on Function

asmeurer opened the pull request:
This lets Python 3's inspect.signature give a useful signature for Function subclasses.

For example,

**BEFORE**

``` py
In [1]: import inspect

In [2]: inspect.signature(LambertW)
Out[2]: <Signature (*args, **options)>
```

**AFTER**

``` py
In [1]: import inspect

In [2]: inspect.signature(LambertW)
Out[2]: <Signature (x, k=None)>
```

I was hoping this would also improve the IPython `?` help, but that is contingent
on https://github.com/ipython/ipython/issues/3678.

asmeurer committed:
Define __signature__ on Function

This lets Python 3's inspect.signature give a useful signature for Function
subclasses.

For example,

BEFORE

In [1]: import inspect

In [2]: inspect.signature(LambertW)
Out[2]: <Signature (*args, **options)>

AFTER

In [1]: import inspect

In [2]: inspect.signature(LambertW)
Out[2]: <Signature (x, k=None)>

I was hoping this would also improve the IPython ? help, but that is contingent
on https://github.com/ipython/ipython/issues/3678.

minrk commented on sympy/core/function.py:
```diff
@@ -138,6 +138,19 @@ def __init__(cls, *args, **kwargs):
+        Function subclasses.
+        """
+        # Python 3 only, but this method is also only ever called in Python
+        # 3.
```
Since IPython includes a backport of inspect.signature on Python 2, this statement isn't true. It will behave better if you wrap this import in 

```python
        try:
            from inspect import signature
        except ImportError:
            return None
```

asmeurer replied:
Fixed.

asmeurer committed:
Don't assume inspect.signature can be imported in __signature__

IPython has a Python 2 backport of __signature__. See
https://github.com/ipython/ipython/pull/9335.

moorepants commented:
+1

moorepants commented:
Travis is showing that master if failing on this commit now.
